core/arch/atomic_exports.h

```diff
@@ -723,7 +723,254 @@ atomic_dec_becomes_zero(volatile int *var)
                                  : "r2")
 #        define SET_IF_NOT_ZERO(flag) SET_FLAG(ne, flag)
 #        define SET_IF_NOT_LESS(flag) SET_FLAG(ge, flag)
-#    endif /* X86/ARM */
+#    elif defined(DR_HOST_RISCV64)
+#        define ATOMIC_1BYTE_READ(addr_src, addr_res)               \
+            do {                                                    \
+                __asm__ __volatile__("lbu %0, 0(%1)\n\t"            \
+                                     "fence iorw,iorw"              \
+                                     :                              \
+                                     : "r"(addr_src), "r"(addr_res) \
+                                     : "memory");                   \
+            } while (0)
+#        define ATOMIC_1BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_4BYTE_WRITE(target, value, hot_patch)   \
+            do {                                               \
+                /* Not currently used to write code */         \
+                ASSERT_CURIOSITY(!hot_patch);                  \
+                /* Assuming little endian. */                  \
+                ASSERT(BYTE_ORDER == LITTLE_ENDIAN);           \
+                /* FIXME i#3544: test the assembly!*/          \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 1(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 2(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 3(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_4BYTE_ALIGNED_WRITE(target, value, hot_patch) \
+            do {                                                     \
+                /* Not currently used to write code */               \
+                ASSERT_CURIOSITY(!hot_patch);                        \
+                __asm__ __volatile__("fence iorw,iorw\n\t"           \
```

Comment:
I see there are fences in almost all of these implementations. How does the fence help? It would only ensure that there's no reordering around it, but not that the read/written data is visible all at once to all threads/processors.

Reply:
For an operation to be atomic, it needs to be visible "all at once". Would that be true even without the fence?

Reply:
Yes, there will not be a partial write even without a fence.
Fence (write-barrier) ensures ordering (as you've mentioned) and pushes out any previous write operations from a write buffer.
I'm not sure which memory_order (speaking with `stdatomics.h` terms) is the ATOMIC_4BYTE_ALIGNED_WRITE supposed to model. If the "relaxed" then we don't need a fence here. If "release" then the fence should be here to push out the writes from the critical section.

Reply:
DR's ATOMIC macros target release-acquire semantics, so not just ensuring "untorn" or non-partial. 
 There is a comment on this somewhere -- here's one at least (should probably be at top of file too, feel free to add there too in this PR to help others): https://github.com/DynamoRIO/dynamorio/blob/master/core/arch/atomic_exports.h#L926

Reply:
Generally, we want to match what `std::atomic` does for `std::memory_order_release` and `std::memory_order_acquire`, or the C11 equivalent.

Reply:
Xref #4264

Reply:
If you could double-check/confirm that all the fences in the PR are appropriate for release/acquire.

Reply:
Confirmed that I have the following semantics:
- `ATOMIC_*_READ/WRITE()` -> acquire/release order.
- The rest follow sequentially consistent order.

Reply:
> The rest follow sequentially consistent order.

Not sure I follow this: what are "the rest"?

Reply:
I mean the rest of the arch-dependent atomic function definitions in this file, that is:
`ATOMIC_INC*()`, `ATOMIC_DEC*()`, `ATOMIC_ADD*()`, `atomic_add_exchange*()`, `atomic_compare_exchange*()`, `atomic_exchange()`, `atomic_*_and_test()` and `atomic_dec_becomes_zero()`.

Reply:
Those should also only need release-acquire, right?  Though for those there may not be any difference.

Reply:
AFAIK they are seq_cst in x86 and ARM? I.e. `ATOMIC_INC_int()` on X86 it uses `lock incl %0`, on AArch64 it uses `ldaxr+stlxr` loop (which is load-acquire + op + store+release) and ARM uses similar loop and a `dmb ish` which is a read/write seq_cst barrier on memory in a single share'ability domain (IIRC what that means exactly is implementation specific).

Reply:
Taking INC+DEC as an example. Yes, if we had INC+DEC follow acquire+release, then the pipeline would have more leeway in running the code out of order. Then again it would mean that only INC->DEC logic is proper for a critical section, which might lead to creating an acquire and release version of INC and DEC. That is a short path to just ditching this header and going for stdatomic. Though one of the reasons of having it might be if we ever need u128 atomic operations. Those might not be intrinsic functions (IIRC on RISC-V they are part of libatomic).

Reply:
The same sequence might provide seq_cst in some cases, but it is necessary for acquire-release: acquire-release needs the `lock` prefix on x86.  On AArch64 for `std::memory_order_acq_rel` for a `fetch_add`, you end up with `ldaxr..stlxr` .  Similarly on ARM you need the `dmb ish`.  So the sequences for a read-modify-write are similar for acq_rel and seq_cst for the existing arches but I believe the requirement is acq_rel and not seq_cst so if it's different for RISC-V we would want the more relaxed one.

For using stdatomic that is the issue I referenced before: #4264.

Reply:
I just checked `atomic_fetch_add_explicit()`, `atomic_exchange_explicit()` and `atomic_compare_exchange_strong_explicit()` with both `memory_order_acq_rel` and `memory_order_seq_cst` and both versions are identical for each of the atomic functions (at least in GCC 11.2).

Therefore I think that my approach should be valid. `atomic_compare_exchange*()` follows GCC exactly (since it is a `lr+sc` sequence so I saw no other way of doing it) where as the rest (INC, DEC, etc.) use a simpler implementation. I.e. for `ATOMIC_ADD_int()` I have:
```
amoadd.w.aqrl zero,a1,0(a0)
```
  instead of:
```
fence   iorw,ow
amoadd.w.aq     zero,a1,0(a0)
```
The only difference I can think of is that GCC's version does extra ordering of I/O read/writes before I/O writes with the fence where as my version only orders the type of memory that `var` points to. I don't think the I/O ordering really matters for us, right?
Also my version is one instruction shorter.

Reply:
Agreed, your version seems sufficient to me.